lua/luasnip/util/parser/init.lua

```diff
@@ -7,6 +7,9 @@ local functions = require("luasnip.util.functions")
 local util = require("luasnip.util.util")
 local extend_decorator = require("luasnip.util.extend_decorator")
 
+---@class Parser
+---@field parse_snippet fun(context: table|number|string, body: string, opts: table|nil): table
+---@field parse_snipmate fun(context: table|number|string, body: string, opts: table|nil): table
```

Comment:
Note that class annotations are usually only used when there are fields with values, here it is not really a class but a module with only functions.
Also don't duplicate function signature between here and the docstring above each function, it'd be very annoying to maintain.

Reply:
Thanks for the feedback! I used class indeed in order to restrict the fields you can invoke from a module. Is there a way to the same with a table?

That's very helpful, because the starting point was pretty annoying to get done. I would be happy to implement the changes you want if you don't want to do it yourselves 👍

For the docstrings, I tried not to touch the original code too much, because I am not familiar enough with the code. I was hoping someone more expert would help tidy that up.

If there is a better way than submitting PRs to improve the type hints incrementally, I would be happy to follow. But I don't think we can avoid making those kind of roundtrips, because I am too much of a LuaSnoob for understanding what's going on in the code base.

You can count on me however to produce some starting point and do the chores, as it's time-consuming. 

I hope my contributions can be useful to you! Let me know how I can help 😊